lib/Conversion/TritonGPUToLLVM/ReduceOpToLLVM.cpp

```diff
@@ -144,15 +145,29 @@ struct ReduceOpConversion
     triton::ReduceOp op = helper.getOperation();
     RankedTensorType operandType = op.getInputTypes()[0];
     // Assumes offsets don't actually depend on type
-    SmallVector<SmallVector<unsigned>> offset =
+    SmallVector<SmallVector<unsigned>> offsets =
         emitOffsetForLayout(helper.getSrcLayout(), operandType);
+
+    // Thread X might hold the same input value in two registers.  Get the
+    // indices in `offsets` that hold unique values, and only accumualte over
+    // those.
+    //
+    // TODO(jlebar): This issue exists not just for the register level, but for
+    // every level of the hierarchy.  For example, registers in two different
+    // threads can also contain the same value.  As far as I can tell, this code
+    // does not take that into account anywhere.
+    llvm::MapVector<ArrayRef<unsigned>, int> uniqueOffsets;
+    for (int i = 0; i < offsets.size(); ++i) {
+      uniqueOffsets.insert({offsets[i], i});
+    }
```

Comment:
This is handled in the rest of the file by using helper such as `getWarpsPerCTAWithUniqueData`. Is it possible to do the same here? And is there really something to address in the other levels?

Reply:
Aha, thanks.  There does not appear to be getElementsPerThreadWithUniqueData, alas.

Reply:
> There does not appear to be getElementsPerThreadWithUniqueData, alas.

I'm now I'm wondering if we should have such an API/helper. I think it would always equal to `std::min(elementPerThread[dim], shape[dim])`. Are there any other cases where this is not true?

I expect this to happen in more places than here.

Reply:
It's entirely possible that other places are affected, but tests are passing and I'm not sure how to find them except to audit all the code?

> I think it would always equal to std::min(elementPerThread[dim], shape[dim])

I don't think that's 100% true in all edge cases.  For example if you have a 16xf32 tensor distributed among 2 CTAs, each CTA gets half of the tensor, always.  (At least, this is what I have found when writing this PR.)  So each CTA gets 8 unique elements.  Now if I have elemsPerThread = 16, then the registers will hold 2 copies of each value, so uniqueOffsets should have size 8.  But min(elemPerThread[dim], shape[dim]) = 16.

Reply:
Interesting, if that's true I think it breaks all the other helpers as they rely on the same logic (that things gets distributed in order, threads, then warps, then block) :(